Add support for Mariner Linux on Azure

mboersma opened the pull request:
What this PR does / why we need it:

Adds VHD and SIG build support for [Mariner](https://en.wikipedia.org/wiki/CBL-Mariner) Linux in Azure. See documentation at https://microsoft.github.io/CBL-Mariner/docs/.

Which issue(s) this PR fixes:

Fixes #1071

**Additional context**

I think we should scope this to just Azure support, then follow up with OVA.

mboersma commented:
This builds a VHD successfully, but when I make a managed image from that and use it in Cluster API, `kubeadm init` never happens. Any ideas or debugging help would be very welcome!

ader1990 commented:
> This builds a VHD successfully, but when I make a managed image from that and use it in Cluster API, `kubeadm init` never happens. Any ideas or debugging help would be very welcome!

Hello, most likely, cloud-init did not run or did run incorrectly. I think it is needed to make sure that the image has been maybe generalized (aka cleaned from previous runs of cloud-init)?

Cloud-init logs are usually in the two files -> /var/log/cloud*.

mboersma commented on images/capi/Makefile:
```diff
@@ -625,6 +627,7 @@ build-azure-vhd-ubuntu-1804: ## Builds Ubuntu 18.04 VHD image for Azure
 build-azure-vhd-ubuntu-2004: ## Builds Ubuntu 20.04 VHD image for Azure
 build-azure-vhd-ubuntu-2204: ## Builds Ubuntu 22.04 VHD image for Azure
 build-azure-vhd-centos-7: ## Builds CentOS 7 VHD image for Azure
+build-azure-vhd-mariner-2: ## Builds Mariners 2 VHD image for Azure
```
I'm planning to add it in this one, once I work out the init issues.

ader1990 commented on images/capi/packer/ova/mariner-2.json:
```diff
@@ -0,0 +1,16 @@
+  "guest_os_type": "generic-linux",
+  "iso_checksum": "a32238f58f68b348f136c54620bb3a3c60b446ec21fdb9185ecd8979c0fc6e8b",
+  "iso_checksum_type": "sha256",
+  "iso_url": "https://aka.ms/mariner-2.0-x86_64-iso",
```
Hello,

I have started with this PR and managed to create a Mariner 2 image using QEMU-KVM, with the ISO provided by MSFT. Besides the manual steps I needed to install from ISO (mostly enter + tab that can be eventually automated using boot_command_prefix), the resulting install needs a few more steps to be executed:

 - install cloud-init and jq packages
 - disable firewall (by default only port 22 is allowed in /etc/systemd/scripts/ip4save).
    A K8s node needs port 6443 and other ports allowed, depeding on the services / workloads installed
 - user builder with password builder created | enabled with no sudo pass: echo "builder ALL=(ALL:ALL) NOPASSWD:ALL" >> /etc/sudoers

  Tested the image with CAPI and CAPT (cluster api provider for Tinkerbell) and worked as expected.

mboersma replied:
@ader1990 that's great, thanks so much for testing!

I had only stubbed out the boot_command_prefix but I'm planning to flesh it out once I have the other issues fixed. I'm working on opening port 6443 in iptables and hopefully I can address the other things on your list at the same time.

ader1990 replied:
We have a WiP for the boot command here, should we make a PR on your branch?
https://github.com/mpanduru/image-builder/commit/7220c82dda6e196956a80d3d19713a3a8e012fe5

mboersma commented:
After figuring out how to open the kube-apiserver port persistently through iptables, this seems to be working, at least for my testing through `tilt` and a customized template pointing to my managed image. ❗ 

I'll do some more testing and clean it up, then look at the -sig and ISO targets.

mboersma commented:
One thing I can't figure out is why the mariner.yml node task is run twice:

```
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [include_role : node] *****************************************************
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : perform a yum update] ********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install baseline dependencies on Mariner] ************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install extra rpms] **********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : perform a yum update] ********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install baseline dependencies on Mariner] ************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [setup : install extra rpms] **********************************************
    azure-arm.sig-mariner-2: ok: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Ensure overlay module is present] *********************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Ensure br_netfilter module is present] ****************************
    azure-arm.sig-mariner-2: changed: [default]
    azure-arm.sig-mariner-2:
    azure-arm.sig-mariner-2: TASK [node : Persist required kernel modules] **********************************
    azure-arm.sig-mariner-2: changed: [default]
```

I can't see any circular import-type thing, and I'm not even sure if I introduced this. It's benign because the tasks are designed to be idempotent, but shouldn't be happening. Ideas welcome!

AverageMarcus reviewed (commented):
/lgtm

Only thing I'm not sure about is the naming - is `Mariner 2` the name Azure is publicly using to refer to the OS or is it using `Azure Linux` ?

mboersma commented:
>  is Mariner 2 the name Azure is publicly using to refer to the OS or is it using Azure Linux ?

It's been called "Mariner" publicly, but recently I saw a press release that it's being rebranded to "Azure Linux" so I ran with that. But I'm not seeing many references to "Azure Linux" yet in the docs, and everyone still calls it "Mariner" AFAICT.

I'd be happy just to revert to "Mariner" everywhere and remove mentions of "Azure Linux" if you think that's clearer. I wasn't 100% clear on which way to go.

mboersma commented on images/capi/ansible/roles/kubernetes/tasks/mariner.yml:
```diff
@@ -0,0 +1,46 @@
+  iptables:
+    chain: INPUT
+    protocol: tcp
+    destination_port: 6443
```
Maybe this should be a user var?

And what if a runtime configuration changes the api server listen port from the default 6443? AFAICT the approach [in Photon](https://github.com/kubernetes-sigs/image-builder/blob/main/images/capi/ansible/roles/setup/tasks/photon.yml#L51-L61) is to allow listening on all ports, but that's not very secure.

Also, should this move out of the `kubernetes` role into the `setup` role? That's where other iptables changes seem to be, but I thought this was specific to kube-apiserver so could stay here.

mboersma commented on images/capi/ansible/roles/containerd/tasks/main.yml:
```diff
@@ -16,7 +16,7 @@
 
 - import_tasks: redhat.yml
-  when: ansible_os_family == "RedHat"
+  when: ansible_os_family in ["Common Base Linux Mariner", "RedHat"]
```
Ansible doesn't recognize Mariner specifically, so `ansible_distribution` and `ansible_os_family` are both "Common Base Linux Mariner". I could fabricate a fact so we can refer to it just as "Mariner", like Flatcar does, if that would be an improvement.

AverageMarcus commented:
I'm happy with whichever name is going to be most known in the ecosystem. I was just wondering if this would align more with "Amazon Linux" or not.

mboersma commented:
>  if this would align more with "Amazon Linux" or not.

I'm thinking it might be more appropriate to go with "Mariner" consistently. I only see "Azure Linux" actually being used today to refer to the AKS service-specific images based on Mariner.

AverageMarcus commented:
Happy to stick with Mariner then! :) 

Mind updating the port to use a user var and then I'm happy to give this an approve :)

mboersma commented:
/retitle Add support for Mariner Linux on Azure

mboersma commented:
/hold

I need to verify `LimitNOFILE` as described in [this comment](https://github.com/kubernetes-sigs/image-builder/issues/1071#issuecomment-1518343092).

mboersma commented:
/hold cancel

AverageMarcus commented:
/lgtm
/approve

Nice work! 😄
